framework/src/mfem/solvers/MFEMSolverBase.C

```diff
@@ -14,8 +14,44 @@ MFEMSolverBase::validParams()
 }
 
 MFEMSolverBase::MFEMSolverBase(const InputParameters & parameters)
-  : MFEMGeneralUserObject(parameters), _lor{getParam<bool>("low_order_refined")}
+  : MFEMGeneralUserObject(parameters),
+    _lor{getParam<bool>("low_order_refined")},
+    _solver{nullptr},
+    _preconditioner{nullptr}
 {
 }
 
+template <typename T>
+void
+MFEMSolverBase::setPreconditioner(T & solver)
+{
+  static_assert(std::is_base_of_v<mfem::Solver, T>);
+
+  if (isParamSetByUser("preconditioner"))
+  {
+    if (!_preconditioner)
+      _preconditioner =
+          &const_cast<MFEMSolverBase &>(getUserObject<MFEMSolverBase>("preconditioner"));
+
+    auto & mfem_pre = _preconditioner->getSolver();
+    if constexpr (std::is_base_of_v<mfem::HypreSolver, T>)
+      if (auto * const hypre_pre = dynamic_cast<mfem::HypreSolver *>(&mfem_pre))
+        solver.SetPreconditioner(*hypre_pre);
+      else
+        mooseError("hypre solver preconditioners must themselves be hypre solvers");
+    else if constexpr (std::is_base_of_v<mfem::IterativeSolver, T>)
+      solver.SetPreconditioner(mfem_pre);
+    else
+      mooseAssert(
+          true,
+          "We should never get here because we only instantiate this method for iterative solvers");
```

Comment:
```suggestion
    else
      solver.SetPreconditioner(mfem_pre);
```

Since, as you note, we're explicitly instantiating anyway, maybe we could just get rid of both assertions in this method...? The compiler will complain anyway if `SetPreconditioner` isn't a member of `T`.

The suggestion above should actually cover all cases, but one... (sigh)
`mfem::HypreAME` which takes an `mfem::HypreSolver` for preconditioner, but is not itself an `mfem::HypreSolver`. But again, if we ever implement that solver and instantiate this function, the compiler will complain `mfem_pre` is of the wrong type, at which point we can amend line 37 with `|| T == mfem::HypreAME`.